### ISSUE TITLE
undef *^H is not ambiguous

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12056]
Migrated from [rt.perl.org#112456](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112456) (status was 'resolved')

Searchable as RT112456$

**Comment by p5pRT** [comment]
### From @cpansprout

$ perl5\.15\.9 \-e 'undef \*^H'
Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.



---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.15\.9&#8203;:

Configured by sprout at Tue Apr 10 09&#8203;:26&#8203;:12 PDT 2012\.

Summary of my perl5 \(revision 5 version 15 subversion 9\) configuration&#8203;:
&nbsp;  Snapshot of&#8203;: 2653c1e3b10f71430de1ab8ab0417ca5b048ab19
&nbsp;  Platform&#8203;:
&nbsp;    osname=darwin\, osvers=10\.5\.0\, archname=darwin\-2level
&nbsp;    uname='darwin pint\.local 10\.5\.0 darwin kernel version 10\.5\.0&#8203;: fri nov 5 23&#8203;:20&#8203;:39 pdt 2010; root&#8203;:xnu\-1504\.9\.17~1release\_i386 i386 '
&nbsp;    config\_args='\-de \-Dusedevel \-DDEBUGGING'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;    optimize='\-O3 \-g'\,
&nbsp;    cppflags='\-fno\-common \-DPERL\_DARWIN \-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.2\.1 \(Apple Inc\. build 5664\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib
&nbsp;    libs=\-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;    perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;    libc=\, so=dylib\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.15\.9&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.9/darwin\-2level
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.9
&nbsp;    /usr/local/lib/perl5/5\.15\.9/darwin\-2level
&nbsp;    /usr/local/lib/perl5/5\.15\.9
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl 5\.15\.9&#8203;:
&nbsp;    DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;    HOME=/Users/sprout
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/usr/local/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Apr 15 13&#8203;:37&#8203;:23 2012\, sprout wrote&#8203;:
> $ perl5\.15\.9 \-e 'undef \*^H'
> Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.
> 

Is this the same phenomenon?

\#\#\#\#\#
$ \./perl \-e 'undef \*^O'
Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.
\[perl\] 514 $ \./perl \-\-version

This is perl 5\, version 15\, subversion 9 \(v5\.15\.9
\(v5\.15\.7\-759\-g84cf752\*\)\) built for darwin\-2level
\(with 328 registered patches\, see perl \-V for more detail\)
\.\.\.
\#\#\#\#\#

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 15 15&#8203;:38&#8203;:30 2012\, jkeenan wrote&#8203;:
> On Sun Apr 15 13&#8203;:37&#8203;:23 2012\, sprout wrote&#8203;:
> > $ perl5\.15\.9 \-e 'undef \*^H'
> > Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.
> > 
> 
> Is this the same phenomenon?

I imagine so\.

> 
> \#\#\#\#\#
> $ \./perl \-e 'undef \*^O'
> Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.
> \[perl\] 514 $ \./perl \-\-version
> 
> This is perl 5\, version 15\, subversion 9 \(v5\.15\.9
> \(v5\.15\.7\-759\-g84cf752\*\)\) built for darwin\-2level
> \(with 328 registered patches\, see perl \-V for more detail\)
> \.\.\.
> \#\#\#\#\#

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sun Apr 15 15&#8203;:38&#8203;:30 2012\, jkeenan wrote&#8203;:
> On Sun Apr 15 13&#8203;:37&#8203;:23 2012\, sprout wrote&#8203;:
> > $ perl5\.15\.9 \-e 'undef \*^H'
> > Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.
> > 
> 
> Is this the same phenomenon?

I imagine so\.

> 
> \#\#\#\#\#
> $ \./perl \-e 'undef \*^O'
> Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.
> \[perl\] 514 $ \./perl \-\-version
> 
> This is perl 5\, version 15\, subversion 9 \(v5\.15\.9
> \(v5\.15\.7\-759\-g84cf752\*\)\) built for darwin\-2level
> \(with 328 registered patches\, see perl \-V for more detail\)
> \.\.\.
> \#\#\#\#\#

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @ilmari

This fixes the following bogus warning \[perl \#112456\]&#8203;:

&nbsp;  $ perl \-e 'undef \*^H'
&nbsp;  Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.

Compare to the non\-warning variant&#8203;:

&nbsp;    $ perl \-e 'undef \*\{^H\}'


---
 t/lib/warnings/toke |    2 \+\+
 toke\.c              |    4 \+\+\-\-
 2 files changed\, 4 insertions\(\+\)\, 2 deletions\(\-\)

<details><summary>Inline Patch</summary>

```diff
diff --git a/t/lib/warnings/toke b/t/lib/warnings/toke
index dd8dc3d..a9106fb 100644
--- a/t/lib/warnings/toke
+++ b/t/lib/warnings/toke
@@ -662,6 +662,8 @@ yelp at foo line 30.
 ########
 # toke.c
 my $a = rand + 4 ;
+$a = rand *^H ;
+$a = rand $^H ;
 EXPECT
 Warning: Use of "rand" without parentheses is ambiguous at - line 2.
 ########
diff --git a/toke.c b/toke.c
index 1d18550..6cc0336 100644
--- a/toke.c
+++ b/toke.c
@@ -8892,8 +8892,6 @@ S_scan_ident(pTHX_ register char *s, register const char *send, char *dest, STRL
 	bracket = s;
 	s++;
     }
-    else if (ck_uni)
-	check_uni();
     if (s < send) {
         if (UTF) {
             const STRLEN skip = UTF8SKIP(s);
@@ -8911,6 +8909,8 @@ S_scan_ident(pTHX_ register char *s, register const char *send, char *dest, STRL
 	*d = toCTRL(*s);
 	s++;
     }
+    else if (ck_uni && !bracket)
+	check_uni();
     if (bracket) {
 	if (isSPACE(s[-1])) {
 	    while (s < send) {
-- 
1.7.5.4
```
</details>

**Comment by p5pRT** [comment]
### From @ilmari

"Father Chrysostomos via RT" \<perlbug\-comment@&#8203;perl\.org> writes&#8203;:

> On Sun Apr 15 15&#8203;:38&#8203;:30 2012\, jkeenan wrote&#8203;:
>> On Sun Apr 15 13&#8203;:37&#8203;:23 2012\, sprout wrote&#8203;:
>> > $ perl5\.15\.9 \-e 'undef \*^H'
>> > Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line 1\.
>> > 
>> 
>> Is this the same phenomenon?
>
> I imagine so\.

RT seems to have decided not to forward my mail\, but here's a suggested
patch I added to the ticket the other day\.

**Comment by p5pRT** [comment]
### From @ilmari

<details><summary>ambiguous-ctrl-glob.diff</summary>

```diff
Subject: [perl #112456] Don't warn about "ambiguous without parens" for ctrl-glob
Date: Mon, 16 Apr 2012 01:05:52 +0100
To: perlbug-comment@perl.org
From: Dagfinn Ilmari Mannsåker <ilmari@ilmari.org>

This fixes the following bogus warning [perl #112456]:

  $ perl -e 'undef *^H'
  Warning: Use of "undef" without parentheses is ambiguous at -e line 1.

Compare to the non-warning variant:

    $ perl -e 'undef *{^H}'
---
 t/lib/warnings/toke |    2 ++
 toke.c              |    4 ++--
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/t/lib/warnings/toke b/t/lib/warnings/toke
index dd8dc3d..a9106fb 100644
--- a/t/lib/warnings/toke
+++ b/t/lib/warnings/toke
@@ -662,6 +662,8 @@ yelp at foo line 30.
 ########
 # toke.c
 my $a = rand + 4 ;
+$a = rand *^H ;
+$a = rand $^H ;
 EXPECT
 Warning: Use of "rand" without parentheses is ambiguous at - line 2.
 ########
diff --git a/toke.c b/toke.c
index 1d18550..6cc0336 100644
--- a/toke.c
+++ b/toke.c
@@ -8892,8 +8892,6 @@ S_scan_ident(pTHX_ register char *s, register const char *send, char *dest, STRL
 	bracket = s;
 	s++;
     }
-    else if (ck_uni)
-	check_uni();
     if (s < send) {
         if (UTF) {
             const STRLEN skip = UTF8SKIP(s);
@@ -8911,6 +8909,8 @@ S_scan_ident(pTHX_ register char *s, register const char *send, char *dest, STRL
 	*d = toCTRL(*s);
 	s++;
     }
+    else if (ck_uni && !bracket)
+	check_uni();
     if (bracket) {
 	if (isSPACE(s[-1])) {
 	    while (s < send) {
-- 
1.7.5.4


```
</details>

**Comment by p5pRT** [comment]
### From @ilmari


\-\- 
ilmari
"A disappointingly low fraction of the human race is\,
 at any given time\, on fire\." \- Stig Sandbeck Mathisen

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 15 17&#8203;:06&#8203;:22 2012\, ilmari wrote&#8203;:
> This fixes the following bogus warning \[perl \#112456\]&#8203;:
> 
>   $ perl \-e 'undef \*^H'
>   Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line
> 1\.
> 
> Compare to the non\-warning variant&#8203;:
> 
>     $ perl \-e 'undef \*\{^H\}'
> \-\-\-
>  t/lib/warnings/toke |    2 \+\+
>  toke\.c              |    4 \+\+\-\-
>  2 files changed\, 4 insertions\(\+\)\, 2 deletions\(\-\)
> 
> diff \-\-git a/t/lib/warnings/toke b/t/lib/warnings/toke
> index dd8dc3d\.\.a9106fb 100644
> \-\-\- a/t/lib/warnings/toke
> \+\+\+ b/t/lib/warnings/toke
> @&#8203;@&#8203; \-662\,6 \+662\,8 @&#8203;@&#8203; yelp at foo line 30\.
>  \#\#\#\#\#\#\#\#
>  \# toke\.c
>  my $a = rand \+ 4 ;
> \+$a = rand \*^H ;
> \+$a = rand $^H ;
>  EXPECT
>  Warning&#8203;: Use of "rand" without parentheses is ambiguous at \- line 2\.
>  \#\#\#\#\#\#\#\#
> diff \-\-git a/toke\.c b/toke\.c
> index 1d18550\.\.6cc0336 100644
> \-\-\- a/toke\.c
> \+\+\+ b/toke\.c
> @&#8203;@&#8203; \-8892\,8 \+8892\,6 @&#8203;@&#8203; S\_scan\_ident\(pTHX\_ register char \*s\, register
> const char \*send\, char \*dest\, STRL
>      bracket = s;
>      s\+\+;
>      \}
> \-    else if \(ck\_uni\)
> \-    check\_uni\(\);
>      if \(s \< send\) \{
>          if \(UTF\) \{
>              const STRLEN skip = UTF8SKIP\(s\);
> @&#8203;@&#8203; \-8911\,6 \+8909\,8 @&#8203;@&#8203; S\_scan\_ident\(pTHX\_ register char \*s\, register
> const char \*send\, char \*dest\, STRL
>      \*d = toCTRL\(\*s\);
>      s\+\+;
>      \}
> \+    else if \(ck\_uni && \!bracket\)
> \+    check\_uni\(\);
>      if \(bracket\) \{
>      if \(isSPACE\(s\[\-1\]\)\) \{
>          while \(s \< send\) \{

Thank you\.  This will have to wait till after the 5\.16 code freeze\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sun Apr 15 17&#8203;:06&#8203;:22 2012\, ilmari wrote&#8203;:
> This fixes the following bogus warning \[perl \#112456\]&#8203;:
> 
>   $ perl \-e 'undef \*^H'
>   Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line
> 1\.
> 
> Compare to the non\-warning variant&#8203;:
> 
>     $ perl \-e 'undef \*\{^H\}'
> \-\-\-
>  t/lib/warnings/toke |    2 \+\+
>  toke\.c              |    4 \+\+\-\-
>  2 files changed\, 4 insertions\(\+\)\, 2 deletions\(\-\)
> 
> diff \-\-git a/t/lib/warnings/toke b/t/lib/warnings/toke
> index dd8dc3d\.\.a9106fb 100644
> \-\-\- a/t/lib/warnings/toke
> \+\+\+ b/t/lib/warnings/toke
> @&#8203;@&#8203; \-662\,6 \+662\,8 @&#8203;@&#8203; yelp at foo line 30\.
>  \#\#\#\#\#\#\#\#
>  \# toke\.c
>  my $a = rand \+ 4 ;
> \+$a = rand \*^H ;
> \+$a = rand $^H ;
>  EXPECT
>  Warning&#8203;: Use of "rand" without parentheses is ambiguous at \- line 2\.
>  \#\#\#\#\#\#\#\#
> diff \-\-git a/toke\.c b/toke\.c
> index 1d18550\.\.6cc0336 100644
> \-\-\- a/toke\.c
> \+\+\+ b/toke\.c
> @&#8203;@&#8203; \-8892\,8 \+8892\,6 @&#8203;@&#8203; S\_scan\_ident\(pTHX\_ register char \*s\, register
> const char \*send\, char \*dest\, STRL
>      bracket = s;
>      s\+\+;
>      \}
> \-    else if \(ck\_uni\)
> \-    check\_uni\(\);
>      if \(s \< send\) \{
>          if \(UTF\) \{
>              const STRLEN skip = UTF8SKIP\(s\);
> @&#8203;@&#8203; \-8911\,6 \+8909\,8 @&#8203;@&#8203; S\_scan\_ident\(pTHX\_ register char \*s\, register
> const char \*send\, char \*dest\, STRL
>      \*d = toCTRL\(\*s\);
>      s\+\+;
>      \}
> \+    else if \(ck\_uni && \!bracket\)
> \+    check\_uni\(\);
>      if \(bracket\) \{
>      if \(isSPACE\(s\[\-1\]\)\) \{
>          while \(s \< send\) \{

Thank you\.  This will have to wait till after the 5\.16 code freeze\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 15 17&#8203;:06&#8203;:22 2012\, ilmari wrote&#8203;:
> This fixes the following bogus warning \[perl \#112456\]&#8203;:
> 
>   $ perl \-e 'undef \*^H'
>   Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line
> 1\.
> 
> Compare to the non\-warning variant&#8203;:
> 
>     $ perl \-e 'undef \*\{^H\}'
> \-\-\-
>  t/lib/warnings/toke |    2 \+\+
>  toke\.c              |    4 \+\+\-\-
>  2 files changed\, 4 insertions\(\+\)\, 2 deletions\(\-\)

Thank you\.  Applied as fbdd83da9d18\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Sun Apr 15 17&#8203;:06&#8203;:22 2012\, ilmari wrote&#8203;:
> This fixes the following bogus warning \[perl \#112456\]&#8203;:
> 
>   $ perl \-e 'undef \*^H'
>   Warning&#8203;: Use of "undef" without parentheses is ambiguous at \-e line
> 1\.
> 
> Compare to the non\-warning variant&#8203;:
> 
>     $ perl \-e 'undef \*\{^H\}'
> \-\-\-
>  t/lib/warnings/toke |    2 \+\+
>  toke\.c              |    4 \+\+\-\-
>  2 files changed\, 4 insertions\(\+\)\, 2 deletions\(\-\)

Thank you\.  Applied as fbdd83da9d18\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
